File: frontend/src/utils/hangulKeyboard.ts
/**
 * 두벌식 한글 자판을 영문 상태로 입력한 문자열을 한글로 조합한다.
 *
 * 예: `tjgmldus` → `서희연`
 * 검색어 보정에 쓰므로, 영문 자판 키가 아닌 문자는 그대로 보존한다.
 */

const KEY_TO_JAMO: Record<string, string> = {
  r: 'ㄱ', R: 'ㄲ', s: 'ㄴ', e: 'ㄷ', E: 'ㄸ', f: 'ㄹ', a: 'ㅁ', q: 'ㅂ', Q: 'ㅃ',
  t: 'ㅅ', T: 'ㅆ', d: 'ㅇ', w: 'ㅈ', W: 'ㅉ', c: 'ㅊ', z: 'ㅋ', x: 'ㅌ', v: 'ㅍ', g: 'ㅎ',
  k: 'ㅏ', o: 'ㅐ', i: 'ㅑ', O: 'ㅒ', j: 'ㅓ', p: 'ㅔ', u: 'ㅕ', P: 'ㅖ', h: 'ㅗ',
  y: 'ㅛ', n: 'ㅜ', b: 'ㅠ', m: 'ㅡ', l: 'ㅣ',
}

const INITIAL_INDEX: Record<string, number> = {
  'ㄱ': 0, 'ㄲ': 1, 'ㄴ': 2, 'ㄷ': 3, 'ㄸ': 4, 'ㄹ': 5, 'ㅁ': 6, 'ㅂ': 7, 'ㅃ': 8,
  'ㅅ': 9, 'ㅆ': 10, 'ㅇ': 11, 'ㅈ': 12, 'ㅉ': 13, 'ㅊ': 14, 'ㅋ': 15, 'ㅌ': 16, 'ㅍ': 17, 'ㅎ': 18,
}

const MEDIAL_INDEX: Record<string, number> = {
  'ㅏ': 0, 'ㅐ': 1, 'ㅑ': 2, 'ㅒ': 3, 'ㅓ': 4, 'ㅔ': 5, 'ㅕ': 6, 'ㅖ': 7, 'ㅗ': 8,
  'ㅘ': 9, 'ㅙ': 10, 'ㅚ': 11, 'ㅛ': 12, 'ㅜ': 13, 'ㅝ': 14, 'ㅞ': 15, 'ㅟ': 16,
  'ㅠ': 17, 'ㅡ': 18, 'ㅢ': 19, 'ㅣ': 20,
}

const FINAL_INDEX: Record<string, number> = {
  '': 0, 'ㄱ': 1, 'ㄲ': 2, 'ㄳ': 3, 'ㄴ': 4, 'ㄵ': 5, 'ㄶ': 6, 'ㄷ': 7, 'ㄹ': 8,
  'ㄺ': 9, 'ㄻ': 10, 'ㄼ': 11, 'ㄽ': 12, 'ㄾ': 13, 'ㄿ': 14, 'ㅀ': 15, 'ㅁ': 16,
  'ㅂ': 17, 'ㅄ': 18, 'ㅅ': 19, 'ㅆ': 20, 'ㅇ': 21, 'ㅈ': 22, 'ㅊ': 23, 'ㅋ': 24,
  'ㅌ': 25, 'ㅍ': 26, 'ㅎ': 27,
}

const COMBINED_MEDIAL: Record<string, string> = {
  'ㅗㅏ': 'ㅘ', 'ㅗㅐ': 'ㅙ', 'ㅗㅣ': 'ㅚ',
  'ㅜㅓ': 'ㅝ', 'ㅜㅔ': 'ㅞ', 'ㅜㅣ': 'ㅟ',
  'ㅡㅣ': 'ㅢ',
}

const COMBINED_FINAL: Record<string, string> = {
  'ㄱㅅ': 'ㄳ', 'ㄴㅈ': 'ㄵ', 'ㄴㅎ': 'ㄶ', 'ㄹㄱ': 'ㄺ', 'ㄹㅁ': 'ㄻ',
  'ㄹㅂ': 'ㄼ', 'ㄹㅅ': 'ㄽ', 'ㄹㅌ': 'ㄾ', 'ㄹㅍ': 'ㄿ', 'ㄹㅎ': 'ㅀ', 'ㅂㅅ': 'ㅄ',
}

const SPLIT_FINAL: Record<string, readonly [string, string]> = {
  'ㄳ': ['ㄱ', 'ㅅ'], 'ㄵ': ['ㄴ', 'ㅈ'], 'ㄶ': ['ㄴ', 'ㅎ'], 'ㄺ': ['ㄹ', 'ㄱ'],
  'ㄻ': ['ㄹ', 'ㅁ'], 'ㄼ': ['ㄹ', 'ㅂ'], 'ㄽ': ['ㄹ', 'ㅅ'], 'ㄾ': ['ㄹ', 'ㅌ'],
  'ㄿ': ['ㄹ', 'ㅍ'], 'ㅀ': ['ㄹ', 'ㅎ'], 'ㅄ': ['ㅂ', 'ㅅ'],
}

function isVowel(jamo: string): boolean {
  return jamo in MEDIAL_INDEX
}

function composeSyllable(initial: string, medial: string, final: string): string {
  const initialIndex = INITIAL_INDEX[initial]
  const medialIndex = MEDIAL_INDEX[medial]
  const finalIndex = FINAL_INDEX[final]
  if (initialIndex == null || medialIndex == null || finalIndex == null) {
    return `${initial}${medial}${final}`
  }
  return String.fromCharCode(0xAC00 + ((initialIndex * 21 + medialIndex) * 28) + finalIndex)
}

/**
 * 영문 상태에서 두벌식 자판으로 입력한 텍스트를 한글로 변환한다.
 * 완성되지 않은 자모와 영문/숫자/공백은 유실하지 않고 그대로 반환한다.
 */
export function englishKeyboardToHangul(input: string): string {
  let result = ''
  let initial = ''
  let medial = ''
  let final = ''

  const flush = () => {
    if (initial && medial) result += composeSyllable(initial, medial, final)
    else result += `${initial}${medial}${final}`
    initial = ''
    medial = ''
    final = ''
  }

  for (const char of input) {
    const jamo = KEY_TO_JAMO[char]
    if (!jamo) {
      flush()
      result += char
      continue
    }

    if (isVowel(jamo)) {
      if (!initial) {
        initial = 'ㅇ'
        medial = jamo
      } else if (!medial) {
        medial = jamo
      } else if (final) {
        const split = SPLIT_FINAL[final]
        if (split) {
          result += composeSyllable(initial, medial, split[0])
          initial = split[1]
        } else {
          result += composeSyllable(initial, medial, '')
          initial = final
        }
        medial = jamo
        final = ''
      } else {
        const combined = COMBINED_MEDIAL[`${medial}${jamo}`]
        if (combined) medial = combined
        else {
          result += composeSyllable(initial, medial, '')
          initial = 'ㅇ'
          medial = jamo
        }
      }
      continue
    }

    if (!initial) initial = jamo
    else if (!medial) {
      flush()
      initial = jamo
    } else if (!final) final = jamo
    else {
      const combined = COMBINED_FINAL[`${final}${jamo}`]
      if (combined) final = combined
      else {
        result += composeSyllable(initial, medial, final)
        initial = jamo
        medial = ''
        final = ''
      }
    }
  }

  flush()
  return result
}

/**
 * 검색어를 비교용 후보 목록으로 확장한다.
 *
 * 원문과 두벌식 한글 변환값을 함께 돌려주므로, 영문 자판으로 입력한 한글
 * (`tjgmldus` → `서희연`)과 실제 영문·숫자 값(사번·이메일 등)이 모두 검색된다.
 * 변환은 대소문자에 의존하므로(`R`→`ㄲ`) 소문자화보다 먼저 수행한다.
 */
export function buildSearchTerms(input: string): string[] {
  const raw = input.trim()
  if (!raw) return []
  return [...new Set([
    raw.toLocaleLowerCase('ko'),
    englishKeyboardToHangul(raw).toLocaleLowerCase('ko'),
  ])]
}

/**
 * 검색 대상 값 중 하나라도 확장된 검색어를 포함하는지 확인한다.
 * `values`의 null/undefined 항목은 무시한다.
 */
export function matchesSearchTerms(
  values: readonly (string | null | undefined)[],
  terms: readonly string[],
): boolean {
  if (terms.length === 0) return true
  const haystack = values
    .filter((value): value is string => !!value)
    .map((value) => value.toLocaleLowerCase('ko'))
  return terms.some((term) => haystack.some((value) => value.includes(term)))
}

File: backend/app/core/hangul_keyboard.py
"""두벌식 한글 자판을 영문 상태로 입력한 검색어를 한글로 복원한다.

한글 입력기를 켜지 않고 검색하면 `서희연`이 `tjgmldus`로 입력된다. 서버 검색은
DB의 실제 한글 값과 ILIKE 비교를 하므로 이런 입력은 아무것도 찾지 못한다.
검색 시 원문과 함께 이 변환 결과도 후보로 넣어 두 경우 모두 매칭시킨다.

프런트엔드 `frontend/src/utils/hangulKeyboard.ts`와 같은 규칙을 따른다.
"""
from __future__ import annotations

KEY_TO_JAMO: dict[str, str] = {
    "r": "ㄱ", "R": "ㄲ", "s": "ㄴ", "e": "ㄷ", "E": "ㄸ", "f": "ㄹ", "a": "ㅁ",
    "q": "ㅂ", "Q": "ㅃ", "t": "ㅅ", "T": "ㅆ", "d": "ㅇ", "w": "ㅈ", "W": "ㅉ",
    "c": "ㅊ", "z": "ㅋ", "x": "ㅌ", "v": "ㅍ", "g": "ㅎ",
    "k": "ㅏ", "o": "ㅐ", "i": "ㅑ", "O": "ㅒ", "j": "ㅓ", "p": "ㅔ", "u": "ㅕ",
    "P": "ㅖ", "h": "ㅗ", "y": "ㅛ", "n": "ㅜ", "b": "ㅠ", "m": "ㅡ", "l": "ㅣ",
}

INITIALS = "ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ"
MEDIALS = "ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ"
FINALS = "_ㄱㄲㄳㄴㄵㄶㄷㄹㄺㄻㄼㄽㄾㄿㅀㅁㅂㅄㅅㅆㅇㅈㅊㅋㅌㅍㅎ"

INITIAL_INDEX = {jamo: index for index, jamo in enumerate(INITIALS)}
MEDIAL_INDEX = {jamo: index for index, jamo in enumerate(MEDIALS)}
# 첫 항목은 종성 없음(0)을 뜻하는 자리표시자이므로 빈 문자열로 바꿔 넣는다.
FINAL_INDEX = {("" if jamo == "_" else jamo): index for index, jamo in enumerate(FINALS)}

COMBINED_MEDIAL: dict[str, str] = {
    "ㅗㅏ": "ㅘ", "ㅗㅐ": "ㅙ", "ㅗㅣ": "ㅚ",
    "ㅜㅓ": "ㅝ", "ㅜㅔ": "ㅞ", "ㅜㅣ": "ㅟ",
    "ㅡㅣ": "ㅢ",
}

COMBINED_FINAL: dict[str, str] = {
    "ㄱㅅ": "ㄳ", "ㄴㅈ": "ㄵ", "ㄴㅎ": "ㄶ", "ㄹㄱ": "ㄺ", "ㄹㅁ": "ㄻ",
    "ㄹㅂ": "ㄼ", "ㄹㅅ": "ㄽ", "ㄹㅌ": "ㄾ", "ㄹㅍ": "ㄿ", "ㄹㅎ": "ㅀ",
    "ㅂㅅ": "ㅄ",
}

SPLIT_FINAL: dict[str, tuple[str, str]] = {
    "ㄳ": ("ㄱ", "ㅅ"), "ㄵ": ("ㄴ", "ㅈ"), "ㄶ": ("ㄴ", "ㅎ"), "ㄺ": ("ㄹ", "ㄱ"),
    "ㄻ": ("ㄹ", "ㅁ"), "ㄼ": ("ㄹ", "ㅂ"), "ㄽ": ("ㄹ", "ㅅ"), "ㄾ": ("ㄹ", "ㅌ"),
    "ㄿ": ("ㄹ", "ㅍ"), "ㅀ": ("ㄹ", "ㅎ"), "ㅄ": ("ㅂ", "ㅅ"),
}

HANGUL_BASE = 0xAC00


def _compose(initial: str, medial: str, final: str) -> str:
    """초성·중성·종성을 완성형 음절로 조합한다. 조합 불가면 자모를 그대로 잇는다."""
    initial_index = INITIAL_INDEX.get(initial)
    medial_index = MEDIAL_INDEX.get(medial)
    final_index = FINAL_INDEX.get(final)
    if initial_index is None or medial_index is None or final_index is None:
        return f"{initial}{medial}{final}"
    return chr(HANGUL_BASE + ((initial_index * 21 + medial_index) * 28) + final_index)


def english_keyboard_to_hangul(text: str) -> str:
    """영문 자판으로 입력한 문자열을 한글로 조합한다.

    자판에 없는 문자(숫자·공백·이미 한글인 문자 등)는 그대로 보존하므로
    사번이나 이메일이 섞인 검색어도 손실 없이 통과한다.
    """
    result: list[str] = []
    initial = ""
    medial = ""
    final = ""

    def flush() -> None:
        nonlocal initial, medial, final
        if initial and medial:
            result.append(_compose(initial, medial, final))
        else:
            result.append(f"{initial}{medial}{final}")
        initial = ""
        medial = ""
        final = ""

    for char in text:
        jamo = KEY_TO_JAMO.get(char)
        if jamo is None:
            flush()
            result.append(char)
            continue

        if jamo in MEDIAL_INDEX:
            if not initial:
                initial = "ㅇ"
                medial = jamo
            elif not medial:
                medial = jamo
            elif final:
                # 종성이 다음 음절의 초성으로 넘어가는 실제 입력기 동작을 재현한다.
                split = SPLIT_FINAL.get(final)
                if split:
                    result.append(_compose(initial, medial, split[0]))
                    initial = split[1]
                else:
                    result.append(_compose(initial, medial, ""))
                    initial = final
                medial = jamo
                final = ""
            else:
                combined = COMBINED_MEDIAL.get(f"{medial}{jamo}")
                if combined:
                    medial = combined
                else:
                    result.append(_compose(initial, medial, ""))
                    initial = "ㅇ"
                    medial = jamo
            continue

        if not initial:
            initial = jamo
        elif not medial:
            flush()
            initial = jamo
        elif not final:
            final = jamo
        else:
            combined = COMBINED_FINAL.get(f"{final}{jamo}")
            if combined:
                final = combined
            else:
                result.append(_compose(initial, medial, final))
                initial = jamo
                medial = ""
                final = ""

    flush()
    return "".join(result)


def expand_search_terms(query: str | None) -> list[str]:
    """검색어를 ILIKE 비교용 후보 목록으로 확장한다.

    원문을 항상 먼저 넣으므로 실제 영문·숫자 검색이 자판 변환 때문에 깨지지 않는다.
    변환 결과가 원문과 같으면 중복을 넣지 않는다.
    """
    normalized = (query or "").strip()
    if not normalized:
        return []
    terms = [normalized]
    converted = english_keyboard_to_hangul(normalized)
    if converted and converted != normalized:
        terms.append(converted)
    return terms
